File: Backend/src/models/health_data.py
from pydantic import BaseModel, Field
from typing import Optional, Any
from datetime import datetime
from bson import ObjectId
from pydantic.json_schema import JsonSchemaValue
from pydantic_core import CoreSchema, core_schema

class PyObjectId(str):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid objectid")
        return str(v)

    @classmethod
    def __get_pydantic_core_schema__(
        cls,
        _source_type: Any,
        _handler: Any,
    ) -> CoreSchema:
        return core_schema.json_or_python_schema(
            json_schema=core_schema.str_schema(),
            python_schema=core_schema.union_schema([
                core_schema.str_schema(),
                core_schema.is_instance_schema(ObjectId),
            ]),
            serialization=core_schema.plain_serializer_function_ser_schema(
                lambda x: str(x), return_schema=core_schema.str_schema(),
            ),
        )

class HealthData(BaseModel):
    id: Optional[PyObjectId] = Field(default=None, alias="_id")
    user_id: Optional[str] = None
    HighBP: int = Field(..., ge=0, le=1)
    HighChol: int = Field(..., ge=0, le=1)
    BMI: float = Field(..., ge=0)
    Smoker: int = Field(..., ge=0, le=1)
    Stroke: int = Field(..., ge=0, le=1)
    HeartDiseaseorAttack: int = Field(..., ge=0, le=1)
    PhysActivity: int = Field(..., ge=0, le=1)
    Fruits: int = Field(..., ge=0, le=1)
    Veggies: int = Field(..., ge=0, le=1)
    HvyAlcoholConsump: int = Field(..., ge=0, le=1)
    AnyHealthcare: int = Field(..., ge=0, le=1)
    NoDocbcCost: int = Field(..., ge=0, le=1)
    GenHlth: int = Field(..., ge=1, le=5)
    MentHlth: int = Field(..., ge=0, le=30)
    PhysHlth: int = Field(..., ge=0, le=30)
    DiffWalk: int = Field(..., ge=0, le=1)
    Sex: int = Field(..., ge=0, le=1)
    Age: int = Field(..., ge=18, le=120)
    Education: int = Field(..., ge=1, le=6)
    Income: int = Field(..., ge=1, le=8)
    created_at: datetime = Field(default_factory=lambda: datetime.now())
    updated_at: datetime = Field(default_factory=lambda: datetime.now())

    model_config = {
        "json_encoders": {ObjectId: str},
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_schema_mode": "serialization"
    }

    def update_timestamp(self):
        self.updated_at = datetime.now() 

File: app/lib/Model/HealthData.dart
class HealthData {
  final String id;
  final double bmi;
  final int stroke;
  final int heartDiseaseorAttack;
  final int sex;
  final int age;
  final String riskLevel;
  final double riskProbability;
  final DateTime createdAt;

  HealthData({
    required this.id,
    required this.bmi,
    required this.stroke,
    required this.heartDiseaseorAttack,
    required this.sex,
    required this.age,
    required this.riskLevel,
    required this.riskProbability,
    required this.createdAt,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bmi': bmi,
      'stroke': stroke,
      'heartDiseaseorAttack': heartDiseaseorAttack,
      'sex': sex,
      'age': age,
      'riskLevel': riskLevel,
      'riskProbability': riskProbability,
      'createdAt': createdAt.toIso8601String(),
    };
  }

  factory HealthData.fromJson(Map<String, dynamic> json) {
    return HealthData(
      id: json['id'],
      bmi: json['bmi'].toDouble(),
      stroke: json['stroke'],
      heartDiseaseorAttack: json['heartDiseaseorAttack'],
      sex: json['sex'],
      age: json['age'],
      riskLevel: json['riskLevel'],
      riskProbability: json['riskProbability'].toDouble(),
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}
